constantine/math/polynomials/polynomials.nim

```diff
@@ -48,70 +57,192 @@ type
+  for i in 13 ..< poly.coefs.len:
+    var degree {.noinit.}: Field
+    degree.fromInt(i)
+    polyprime.coefs[i-1].prod(poly.coefs[i], degree)
+
+# Polynomials in evaluation/Lagrange form
+# ------------------------------------------------------
+
+func areStrictlyIncreasing[Field](a: openArray[Field]): bool {.used.} =
+  ## Check whether points in the domain are ordered
+  if a.len == 0:
+    return false
+
+  var prev = a[0].toBig()
+  for i in 1 ..< a.len:
+    let cur = a[i].toBig()
+    if bool(cur <= prev):
+      return false
+  return true
+
+type InvDiffArrayKind* = enum
+  kArrayMinusZ
+  kZminusArray
+
+func inverseDifferenceArrayZ*[N: static int, Field](
+       r: var array[N, Field],
+       w: array[N, Field],
        z: Field,
+       differenceKind: static InvDiffArrayKind,
        earlyReturnOnZero: static bool): int =
-  ## Compute 1/(ωⁱ-z) for i in [0, N)
+  ## Compute 1/(wᵢ-z) or 1/(z-wᵢ) for i in [0, N)
   ##
-  ## Returns -1 if z ∉ {1, ω, ω², ... , ωⁿ⁻¹}
-  ## Returns the index of ωⁱ==z otherwise
+  ## Returns -1 if z ∉ {w₀, ω₁, ω₂, ... , wₙ₋₁}
+  ## Returns the index of wᵢ==z otherwise
   ##
-  ## If ωⁱ-z == 0 AND earlyReturnOnZero is false
+  ## If wᵢ-z == 0 AND earlyReturnOnZero is false
   ##   the other inverses are still computed
   ##   and 0 is returned at that index
-  ## If ωⁱ-z == 0 AND earlyReturnOnZero is true
-  ##   the index of ωⁱ==z is returned
-  ##   the content of invRootsMinusZ is undefined
+  ## If wᵢ-z == 0 AND earlyReturnOnZero is true
+  ##   the index of wᵢ==z is returned
+  ##   the content of the result r is undefined
+  ##
+  ## The wᵢ must be unique
+  ## so that wᵢ == 0 can happen only once.
+  ##
+  ## **variable-time**:
+  ## This leaks whether z is in the domain or not.
 
   # Mongomery's batch inversion
-  # ω is a root of unity of order N,
-  # so if ωⁱ-z == 0, it can only happen in one place
+  # The wᵢ are unique
+  # so if wᵢ-z == 0, it can only happen in one place
   var accInv{.noInit.}: Field
-  var rootsMinusZ{.noInit.}: array[N, Field]
+  var diffs{.noInit.}: array[N, Field]
 
   accInv.setOne()
   var index0 = -1
 
-  when earlyReturnOnZero: # Split computation in 2 phases
-    for i in 0 ..< N:
-      rootsMinusZ[i].diff(domain.rootsOfUnity[i], z)
-      if rootsMinusZ[i].isZero().bool():
-        return i
-
   for i in 0 ..< N:
-    when not earlyReturnOnZero: # Fused substraction and batch inversion
-      rootsMinusZ[i].diff(domain.rootsOfUnity[i], z)
-      if rootsMinusZ[i].isZero().bool():
+    when differenceKind == kArrayMinusZ:
+      diffs[i].diff(w[i], z)
+    else:
+      diffs[i].diff(z, w[i])
+
+    if diffs[i].isZero().bool():
+      when earlyReturnOnZero:
+        return i
+      else:
         index0 = i
-        invRootsMinusZ[i].setZero()
-        continue
 
-    invRootsMinusZ[i] = accInv
-    accInv *= rootsMinusZ[i]
+  for i in 0 ..< N:
+    if i != index0:
+      r[i] = accInv
+      accInv *= diffs[i]
+    else:
+      r[i].setZero()
```

Comment:
`Fused substraction and batch inversion` doesn't seem to improve performance while it hurts readability of the control flow when you have to parse the earlyReturn conditions